src/gsi/core_disp_func.F90

```diff
@@ -0,0 +1,1055 @@
+  use fp_types_m, only: fp
+  use core_disp_types_m, only: fca_gridded_meta, fca_gridded_disp, &
+       ids,ide,jds,jde,kds,kde, ims,ime,jms,jme,kms,kme, &
+       its,ite,jts,jte,kts,kte, ips,ipe,jps,jpe,kps,kpe, &
+       bilinear, bicubic
+  implicit none
+  private
+  public :: bilin_interp, bicub_interp, cubicInterpolate, compute_xy_orig, apply_disp_2d, apply_vert
+  public :: init_meta, allocate_disp, deallocate_disp
+#ifdef DM_PARALLEL
+  public :: compute_xy_orig_dm
+#endif
+  contains
+#else
+#define TRACE_USE
+#endif
+
+real(fp) function bicub_interp(fin,i,j,dx,dy)
+!$$$  subprogram documentation block
+!                .      .    .                                       .
+! subprogram:    bicub_interp
+!   prgmmr: Nehrkorn
+!
+! abstract:
+!
+! program history log:
+!
+!   input argument list:
+!     fin, dx, dy
+!     i, j
+!
+!   output argument list:
+!
+! attributes:
+!   language: f90
+!   machine:
+!
+!$$$ end documentation block
+  implicit none
+  real(fp), intent(in) :: fin(ims:ime,jms:jme), dx, dy
+  integer, intent(in) :: i, j
+
+  real(fp) :: yy(-1:2)
+  integer :: jj
+
+  do jj=-1,2
+     yy(jj) = cubicInterpolate(fin((i-1):(i+2),j+jj),dx)
+  end do
+  bicub_interp = cubicInterpolate(yy,dy)
+end function bicub_interp
+
+real(fp) function cubicInterpolate(f,x)
+!$$$  subprogram documentation block
+!                .      .    .                                       .
+! subprogram:    cubicInterpolate
+!   prgmmr: Nehrkorn
+!
+! abstract:
+!
+! program history log:
+!
+!   input argument list:
+!     f
+!     x
+!
+!   output argument list:
+!
+! attributes:
+!   language: f90
+!   machine:
+!
+!$$$ end documentation block
+  implicit none
+  real(fp), intent(in) :: f(-1:2) !values at x=-1,0,1,2
+  real(fp), intent(in) :: x ! value in [0,1]
+
+!!$    cubicInterpolate = f(0) + 0.5 * &
+!!$         x*(f(1) - f(-1) + &
+!!$         x*(2.0*f(-1) - 5.0*f(0) + 4.0*f(1) - f(2) + &
+!!$         x*(3.0*(f(0) - f(1)) + f(2) - f(-1))))
+
+  cubicInterpolate = f(0) + &
+       x     * 0.5 * (f(1) - f(-1)) + &
+       x*x   * 0.5 * (2.0*f(-1) - 5.0*f(0) + 4.0*f(1) - f(2)) + &
+       x*x*x * 0.5 * (3.0*(f(0) - f(1)) + f(2) - f(-1))
+
+end function cubicInterpolate
+
+real(fp) function bilin_interp(f00, f10, f01, f11, dx, dy)
+!$$$  subprogram documentation block
+!                .      .    .                                       .
+! subprogram:    bilin_interp
+!   prgmmr: Nehrkorn
+!
+! abstract: bilinear interpolation of 4 nearest neighbors to a point
+!
+! program history log:
+!
+!   input argument list:
+!     f00, f10, f01, f11, dx, dy
+!
+!   output argument list:
+!
+! attributes:
+!   language: f90
+!   machine:
+!
+!$$$ end documentation block
+    ! bilinear interpolation of 4 nearest neighbors to a point
+    implicit none
+    ! bilinear interpolate in the unit box.
+    real(fp), intent(in) :: f00, f10, f01, f11, dx, dy
+    bilin_interp = f00*((1.-dx)*(1.-dy)) + f10*(dx*(1.-dy)) + f01*((1.-dx)*dy) + f11*(dx*dy)
```

Comment:
Please add proper precision to real values (i.e., 1._r_kind)

Reply:
Done.